### DISCUSSION TITLE
Renaming of subclass solvers

### DISCUSSION THREAD

**Discussion by Naktakala** [chi-tech/chi-tech#260] [General]
Quoted from @zachhardy:
" there is inconsistency in subclass naming across the different transport solvers. The first thing is that LBS<solver_type>Solver is redundant and should be LB<solver_type>Solver. I would recommend renaming to a consistent convention. I kind of like LBS<solver_type> because it is shorter. With this change, I would recommend a file renaming to lbs_<solver_name>_<operation> for those classes affected.

Thoughts?"
@Naktakala @ragusa

**Comment by Naktakala** [comment]
Here's what we have now:
- `lbs::SteadySolver`
- `lbs::TransientSolver `
- `lbs_curvilinear::Solver`
- `lbs_adjoint::AdjointSolver`
- `lbs::KEigenvalueSolver`

**Comment by ragusa** [comment]
For acronyms, 3 letters seem to sound good and we have been saying LBS for a couple of years now, so I am accustomed to it, but I see the point. My preference would be to keep the S in LBS. (we can even say that S is for system…)

Now, for the names, I propose those changes:

  *   lbs_curvilinear::SteadySolver
  *   lbs_adjoint::SteadySolver

Thoughts??


--
Jean C. Ragusa, Ph.D.
Professor, Department of Nuclear Engineering
Associate Director, Institute for Scientific Computation
Texas A&M University
Email: ***@***.******@***.***>

Websites:
http://multiphysics.engr.tamu.edu
https://nustem.engr.tamu.edu/

College Station, TX 77843-3133
Phone: (979) 862-2033
FAX: (979) 845-6443
- - - - - - - - - - - - - - - - - - - - - - - -
TEXAS A&M UNIVERSITY | FEARLESS on Every Front

On Dec 23, 2022, at 8:46 AM, Doctor Jan ***@***.******@***.***>> wrote:

This Message Is From an External Sender
This message came from outside your organization.


Here's what we have now:

  *   lbs::SteadySolver
  *   lbs::TransientSolver
  *   lbs_curvilinear::Solver
  *   lbs_adjoint::AdjointSolver
  *   lbs::KEigenvalueSolver

**Reply by zachhardy** [reply]
I second this

**Reply by Naktakala** [reply]
Me too. So this handles class naming.

**Comment by zachhardy** [comment]
My primary concern is the naming convention of the directories and files. The way it looks to me is that we have ''groups'' of solvers: Cartesian, curvilinear, and adjoint. This would naturally lead to the names `LBSSteady` (or `LBSCartesianSteady`), `LBSCurvilinearSteady`, and `LBSAdjointSteady` with the addition of the KEigenvalue and Transient flavors. To me, a directory structure containing these modules makes the most sense. For example, the Cartesian steady solver could be housed under `LinearBoltzmannSolvers/Cartesian/SteadySolver`. We could then make the file naming convention something along the lines of `lbs_<dir1>_<dir2>_<filename>`. This would naturally lead to the namespace structure `lbs::cartesian::SteadySolver`, and so on.

This is a superficial issue, obviously, but I think having a set structure for modules will be better for the code base overall as it grows. My proposal is `<equation_type>/<subcategory1>/<subcategory2>/...` with files named `<equation_type>_<subcategory1>_<subcategory2>_ ... _<filename>`.

Thoughts?

**Reply by Naktakala** [reply]
## Folders in modules
This might be splitting it too far. I've certainly tried different classifications, e.g., sub-categorizing by spatial discretization for which I had FV, PWL, PWLC, PWLCurvilinear. In the end it just felt like another layer of the russian doll that had to be explained to other people.

I would opt not to put another level of folders in here at this moment as a matter of convention purely because it would limit the flexibility to users down the line. My "recommendation" is always to write a solver for a specific discretization but I am not willing to make that a concrete convention because maybe someone wants something different. Silly hypothetical example: someone makes a magneto hydrodynamics solver, and it made sense for them to call the solver some exotic name like `XMagneto`. They figured everything out for 2 or 3 different spatial discretizations and they even have different coordinates systems neatly tucked in. It wouldn't be so nice if we then ask them to go split off their solver into a cartesian and curvilinear coordinate system pieces. If they would like to … then they can. I would rather opt to allow for greater flexibility.

## Filenames in modules
The filenames share some issues with the folders and here too I would prefer not to adopt conventions too harshly. The "Jan style" of lbs_00_constrdestr.cc, lbs_01_initialize, etc. can be considered just that … "my style". There are definitely people that dislike it. But Mr.Magneto up there might be a vim user and likes everything in a single source and header xmagneto.h and xmagneto.cc.

## Folders and Filenames in the ChiTech folder
Now here we can get REAL hardcore on conventions. We open a discussion for ideas on this.

**Reply by zachhardy** [reply]
Jan and I discussed this a few minutes ago and agreed that there needs to be a trade-off between rigidity and flexibility. The preliminary agreement was that as code matures, it should be reorganized to reflect any obvious structure/inheritance.

Tentatively, we agreed on adopting a directory structure that separates modules base on the type of model (governing equations).
We also agreed that for related modules, we should adopt a consistent naming convention that reflects the functionality/specialization of each. For LBS, this is achieved by naming the modules `LBS<specialization>`. This would lead to the directory structure:

- `ChiModules`
    - `LinearBoltzmannSolvers`
        -  `LBSSteady(State)`
        - `LBSKEigenvalue`
        - `LBSTransient`
        - `LBSAdjoint`, `LBSAdjointSteady(State)`, `LBSSteady(State)Adjoint`
        - `LBSCurvilinear`, `LBSCurvilinearSteady(State)`, `LBSSteady(State)Curvilinear`
    - `DiffusionSolvers`
        - `DiffusionSolver`
        - `CFEMDiffusion(Solver)`, `DiffusionCFEM(Solver)`
        -  `DFEMDiffusion(Solver)`, `DiffusionDFEM(Solver`

Note that `(...)` denotes optional. With this structure, modules are grouped by functionality and similarity. New ''base'' modules should be placed either directly in the `ChiModules` folder or into a new ''model'' folder within `ChiModules`. For  example, if someone implemented a finite volume hydrodynamics solver, it would be logical to place it under `ChiModules/HydrodynamicsSolvers/<modulename>`.
